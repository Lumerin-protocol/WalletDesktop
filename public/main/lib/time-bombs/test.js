const timeBombs = require('.')

const start = Date.now()

const bomb1 = timeBombs.create(250)

bomb1.onBoom(function () {
  console.log('Should show ~550ms:', Date.now() - start)
})

setTimeout(function () {
  bomb1.reset(200)
  setTimeout(function () {
    bomb1.reset(200)
  }, 150)
}, 200)

const bomb2 = timeBombs.create(100)

bomb2.onBoom(function () {
  console.log('Should show ~350ms:', Date.now() - start)
})

setTimeout(function () {
  bomb2.reset(300)
}, 50)
