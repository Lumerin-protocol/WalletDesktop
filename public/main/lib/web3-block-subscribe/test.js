const url = 'ws://parity.bloqrock.net:8546'
// const url = 'ws://localhost:8546'

const web3Subscription = require('.')

web3Subscription.subscribe({
  url,
  onData: function ({ number }) {
    console.log('Hash', number)
  },
  onError: function (err) {
    console.warn('Error', err.message)
  }
})
