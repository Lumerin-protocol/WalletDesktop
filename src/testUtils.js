/* eslint-disable require-path-exists/exists */
/* eslint-disable import/no-unresolved */
import { Provider as ClientProvider } from 'lumerin-wallet-ui-logic/src/hocs/clientContext';
import { Provider, createStore } from 'lumerin-wallet-ui-logic/src/store';
import { render, Simulate } from 'react-testing-library';
import { ThemeProvider } from 'styled-components';
import { MemoryRouter } from 'react-router';
import createClient from './client';
import { merge } from 'lodash';
import theme from 'lumerin-wallet-ui-logic/src/theme';
import React from 'react';

import config from '../config';

/*
 * The same render method of 'react-testing-library' but wrapped with
 * Redux store Provider and 'styled-components' ThemeProvider.
 *
 * element      : the React element you want to wrap
 * initialState : optional state to initialize the Redux store
 *
 * Returns the same object as render() but with an additional 'store'
 * property useful for dispatching actions inside your tests.
 */
export function reduxRender(element, initialState) {
  const client = createClient(config, createStore, initialState);

  const renderResult = render(
    <ClientProvider value={client}>
      <Provider store={client.store}>
        <ThemeProvider theme={theme}>
          <React.Fragment>{element}</React.Fragment>
        </ThemeProvider>
      </Provider>
    </ClientProvider>
  );
  return { ...renderResult, store: client.store };
}

/*
 * The same as reduxRender() above but also wrapped with 'react-router'
 * in order to test navigation flows.
 *
 * Returns the same object as reduxRender()
 */
export function routerRender(element, initialState, initialRoute = '/') {
  return reduxRender(
    <MemoryRouter initialEntries={[initialRoute]}>{element}</MemoryRouter>,
    initialState
  );
}

/*
 * Declaratively test complex form validations
 *
 * getter   : function that receives a data-testid and returns an HTML element
 * formId   : data-testid of the form element
 * formData : object of type { [data-testid]: any } to populate the form
 * errors   : object of type { [data-testid]: String } with expected errors
 */
export function testValidation(getter, formId, { formData, errors }) {
  // Populate the specified form fields
  Object.keys(formData).forEach((fieldId, i) => {
    const field = getter(fieldId);
    field.value = formData[fieldId];
    Simulate.change(field);
  });

  // Submit the form
  Simulate.submit(getter(formId));

  // Check expected errors
  Object.keys(errors).forEach((fieldId, i) => {
    expect(getter(`${fieldId}-error`).textContent).toEqual(
      expect.stringContaining(errors[fieldId])
    );
  });
}

/*
 * Add dataset property (data-foo attribute values) to nodes
 *
 * This hack is required because jsdom has no support for dataset yet
 * which prevents us from testing components that rely on it.
 * See https://github.com/jsdom/jsdom/issues/961
 *
 * element   : DOM node to add the dataset to
 * dataAttrs : attribute names to add to the node
 *
 * Usage:
 * <button data-testid="btn" data-modal="buy" data-foo="bar">Click</button>
 *
 * const btn = withDataset(getByTestId('btn'), 'modal', 'foo')
 * // btn.dataset === { modal: 'buy', foo: 'bar'}
 *
 */
export function withDataset(element, ...dataAttrs) {
  element.dataset = dataAttrs.reduce((acc, attr) => {
    acc[attr] = element.getAttribute(`data-${attr}`);
    return acc;
  }, {});
  return element;
}

/**
 * Return UTC timestamps relative to the current time
 * Useful for setting up different auction scenarios.
 */
export const twoWeeksAgo = () => Date.now() / 1000 - 60 * 60 * 24 * 7 * 2;
export const oneHourAgo = () => Date.now() / 1000 - 60 * 60;
export const inOneHour = () => Date.now() / 1000 + 60 * 60;
export const inOneWeek = () => Date.now() / 1000 + 60 * 60 * 24 * 7;

/*
 * Returns a common initial state for Redux store that is useful for most tests
 * Accepts an optional object with overrides that will be deeply merged with
 * the base state object.
 */
export function getInitialState(overrides = {}) {
  const baseState = {
    config,
    connectivity: { isOnline: true },
    blockchain: { height: 1, gasPrice: config.DEFAULT_GAS_PRICE },
    converter: {
      status: {
        availableEth: '100',
        availableLmr: '100',
        currentPrice: '10'
      }
    },
    auction: {
      status: {
        nextAuctionStartTime: inOneHour(),
        tokenRemaining: '1',
        currentAuction: 10,
        currentPrice: '33000000000',
        genesisTime: twoWeeksAgo()
      }
    },
    session: { isLoggedIn: true },
    rates: { ETH: { token: 'ETH', price: 1 } },
    wallets: {
      syncStatus: 'up-to-date',
      active: 'foo',
      allIds: ['foo'],
      byId: {
        foo: {
          addresses: {
            '0x15dd2028C976beaA6668E286b496A518F457b5Cf': {
              token: {
                [config.LMR_TOKEN_ADDR]: { balance: '5000000000000000000000' }
              },
              balance: '5000000000000000000000',
              transactions: []
            }
          }
        }
      }
    }
  };
  return merge({}, baseState, overrides);
}

/*
 * A sample transaction object. Useful for populating state in tests.
 */
export function getDummyTransaction() {
  return {
    meta: {
      lumerin: {},
      tokens: {
        [config.LMR_TOKEN_ADDR]: {
          event: 'Transfer',
          processing: false,
          value:
            '0x000000000000000000000000000000000000000000000001e5b8fa8fe2ac0000'
        }
      }
    },
    receipt: {
      blockHash: '0x1234567890abcdef',
      blockNumber: 407574,
      cumulativeGasUsed: 38484,
      gasUsed: 38484,
      transactionHash: '0x1234567890abcdef',
      transactionIndex: 0
    },
    transaction: {
      transactionIndex: 0,
      blockNumber: 407574,
      blockHash: '0x1234567890abcdef',
      hash: '0x1234567890abcdef',
      value: '0',
      gasPrice: '5000000000',
      gas: 76968,
      from: '0x15dd2028C976beaA6668E286b496A518F457b5Cf',
      to: '0xde806D6efD432CDeE42573760682D99eDEdC1d89'
    }
  };
}
