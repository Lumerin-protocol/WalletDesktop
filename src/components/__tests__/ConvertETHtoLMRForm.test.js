/* eslint-disable require-path-exists/exists */
/* eslint-disable import/no-unresolved */
import ConvertETHtoLMRForm from '../converter/ConvertETHtoLMRForm';
import * as amountFields from './AmountFields.test.js';
import * as gasEditor from './GasEditor.test.js';
import * as testUtils from '../../testUtils';
import { Simulate } from 'react-testing-library';
import React from 'react';

const element = <ConvertETHtoLMRForm tabs={<div />} />;

const ETHprice = 250;

describe('<ConvertETHtoLMRForm/>', () => {
  it.skip('should match its snapshot', () => {
    const { container } = testUtils.reduxRender(element, getInitialState());
    expect(container).toMatchSnapshot();
  });

  describe('When editing the amount fields', () => {
    amountFields.runEditTests(element, getInitialState(), ETHprice);
  });

  describe('When submitting the form', () => {
    amountFields.runValidateTests(element, getInitialState(), 'ethToLmr-form');

    gasEditor.runValidateTests(element, getInitialState(), 'ethToLmr-form');

    it('displays the confirmation view if there are no errors', () => {
      const { queryByTestId, getByTestId } = testUtils.reduxRender(
        element,
        getInitialState()
      );
      expect(queryByTestId('confirmation')).toBeNull();
      const amountField = getByTestId('ethAmount-field');
      const useMinimum = getByTestId('useMinimum-cb');
      amountField.value = '1';
      useMinimum.checked = false;
      Simulate.change(amountField);
      Simulate.change(useMinimum);
      Simulate.submit(getByTestId('ethToLmr-form'));
      expect(queryByTestId('confirmation')).not.toBeNull();
    });
  });
});

function getInitialState() {
  return testUtils.getInitialState({
    rates: { ETH: { token: 'ETH', price: ETHprice } }
  });
}
