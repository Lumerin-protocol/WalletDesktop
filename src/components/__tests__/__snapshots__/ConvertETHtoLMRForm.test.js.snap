// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AmountFields/> should match its snapshot 1`] = `
<div>
  <div
    class="sc-jzJRlG bHelQC"
  >
    <div
      class="sc-fjdhpX drJUfP"
    >
      <button
        class="sc-htpNat fpAZyb"
        data-testid="max-btn"
        tabindex="-1"
        type="button"
      >
        MAX
      </button>
      <div>
        <label
          class="sc-gqjmRU dJlQgb"
          for="ethAmount"
        >
          Amount (ETH)
        </label>
        <input
          class="sc-VigVT bvGVcQ"
          data-testid="ethAmount-field"
          id="ethAmount"
          placeholder="0.00"
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      class="sc-kAzzGY djrWJo"
    >
      <svg
        class="sc-bxivhb hXQRLM"
        viewBox="0 0 24 24"
      >
        <path
          d="M24 7.136v.96H1.642l4.464 4.464-.682.682-5.28-5.28a.475.475 0 0 1 0-.682L5.424 2l.682.682-4.464 4.464H24v-.01zM18.576 11.6l-.682.682 4.464 4.464H0v.96h22.358l-4.464 4.464.682.681 5.28-5.28a.475.475 0 0 0 0-.681l-5.28-5.29z"
        />
      </svg>
    </div>
    <div
      class="sc-fjdhpX drJUfP"
    >
      <div>
        <label
          class="sc-gqjmRU dJlQgb"
          for="usdAmount"
        >
          Amount (USD)
        </label>
        <input
          class="sc-VigVT bvGVcQ"
          data-testid="usdAmount-field"
          id="usdAmount"
          placeholder="0.00"
          type="text"
          value=""
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`<ConvertETHtoLMRForm/> should match its snapshot 1`] = `
<div>
  <div
    class="sc-eNQAEJ hdurla"
  >
    <div
      class="sc-hMqMXs dKMJsu"
    >
      <form
        data-testid="ethToLmr-form"
        novalidate=""
        id="convertForm"
      >
        <div
          class="sc-jKJlTe fkNCah"
        >
          <div
            class="sc-ckVGcZ kvHIve"
          >
            <button
              type="button"
              class="sc-htpNat fpAZyb"
              data-testid="max-btn"
              tabindex="-1"
            >
              MAX
            </button>
            <div>
              <label
                class="sc-dnqmqq ilsafR"
                for="ethAmount"
              >
                Amount (ETH)
              </label>
              <input
                class="sc-iwsKbI kZiHEX"
                type="text"
                id="ethAmount"
                data-testid="ethAmount-field"
                placeholder="0.00"
                value=""
              />
            </div>
          </div>
          <div
            class="sc-hMqMXs dnTKXF"
          >
            <svg
              class="sc-bxivhb hXQRLM"
              viewBox="0 0 24 24"
            >
              <path
                d="M24 7.136v.96H1.642l4.464 4.464-.682.682-5.28-5.28a.475.475 0 0 1 0-.682L5.424 2l.682.682-4.464 4.464H24v-.01zM18.576 11.6l-.682.682 4.464 4.464H0v.96h22.358l-4.464 4.464.682.681 5.28-5.28a.475.475 0 0 0 0-.681l-5.28-5.29z"
              />
            </svg>
          </div>
          <div
            class="sc-ckVGcZ kvHIve"
          >
            <div>
              <label
                class="sc-dnqmqq ilsafR"
                for="usdAmount"
              >
                Amount (USD)
              </label>
              <input
                class="sc-iwsKbI kZiHEX"
                type="text"
                id="usdAmount"
                data-testid="usdAmount-field"
                placeholder="0.00"
                value=""
              />
            </div>
          </div>
        </div>
        <div
          class="sc-hMqMXs bQjXee"
        >
          <div
            class="sc-jKJlTe kWLJKD"
          >
            <div
              class="sc-jKJlTe jGctBw"
            >
              <span
                class="sc-csuQGl izFJpS"
              >
                Gas Limit: 
                250000
                (Units)
              </span>
              <span
                class="sc-csuQGl izFJpS"
              >
                Gas Price: 
                1
                (Gwei)
              </span>
            </div>
            <button
              type="button"
              class="sc-htpNat bqwpzm"
              data-testid="edit-gas-btn"
              tabindex="-1"
            >
              EDIT GAS
            </button>
          </div>
        </div>
        <div
          class="sc-cMljjf jsYJaT"
        >
          Enter a valid amount to get a conversion estimate.
        </div>
        <div>
          <label
            class="sc-jDwBTQ eiQZdA"
          >
            <input
              class="sc-gPEVay hXTYYH"
              data-testid="useMinimum-cb"
              type="checkbox"
              id="useMinimum"
            />
            <span
              class="sc-iRbamj cArvmr"
            >
              Get expected LMR amount or cancel
            </span>
            <span
              class="sc-jlyJG cDmJhB"
              data-rh-darker="true"
              data-rh-width="400px"
              data-rh="This option will cancel the conversion if there is a change in price after you submit the transaction and the expected conversion amount is not lmr.â€¨ Gas will be consumed regardless."
            >
              <svg
                width="14px"
                height="14px"
                viewBox="0 0 14 14"
              >
                <path
                  stroke="none"
                  fill="white"
                  d="M6.33333333,10.3333333 L7.66666667,10.3333333 L7.66666667,6.33333333 L6.33333333,6.33333333 L6.33333333,10.3333333 L6.33333333,10.3333333 Z M7,0.333333333 C3.32,0.333333333 0.333333333,3.32 0.333333333,7 C0.333333333,10.68 3.32,13.6666667 7,13.6666667 C10.68,13.6666667 13.6666667,10.68 13.6666667,7 C13.6666667,3.32 10.68,0.333333333 7,0.333333333 L7,0.333333333 Z M7,12.3333333 C4.06,12.3333333 1.66666667,9.94 1.66666667,7 C1.66666667,4.06 4.06,1.66666667 7,1.66666667 C9.94,1.66666667 12.3333333,4.06 12.3333333,7 C12.3333333,9.94 9.94,12.3333333 7,12.3333333 L7,12.3333333 Z M6.33333333,5 L7.66666667,5 L7.66666667,3.66666667 L6.33333333,3.66666667 L6.33333333,5 L6.33333333,5 Z"
                />
              </svg>
            </span>
          </label>
        </div>
      </form>
    </div>
    <div
      class="sc-bRBYWo ebOYwZ"
    >
      <button
        type="submit"
        class="sc-bwzfXH gAMukr"
        form="convertForm"
      >
        Review Convert
      </button>
    </div>
  </div>
</div>
`;

exports[`<GasEditor/> should match its snapshot 1`] = `
<div>
  <div
    class="sc-jzJRlG bHelQC"
  >
    <div
      class="sc-fjdhpX drJUfP"
    >
      <div>
        <label
          class="sc-gqjmRU dJlQgb"
          for="gasLimit"
        >
          Gas Limit (Units)
        </label>
        <input
          class="sc-VigVT bvGVcQ"
          data-testid="gas-limit-field"
          id="gasLimit"
          type="text"
          value="21000"
        />
      </div>
    </div>
    <div
      class="sc-kAzzGY fJFJon"
    />
    <div
      class="sc-fjdhpX drJUfP"
    >
      <div>
        <label
          class="sc-gqjmRU dJlQgb"
          for="gasPrice"
        >
          Gas Price (Gwei)
        </label>
        <input
          class="sc-VigVT bvGVcQ"
          data-testid="gas-price-field"
          id="gasPrice"
          type="text"
          value="1"
        />
      </div>
    </div>
  </div>
</div>
`;
